vivaldi: add support for proprietary codecs

lluchs opened the pull request:
###### Motivation for this change

As described in #26413, Vivaldi currently doesn't support proprietary codecs such as H.264. This PR adds a `vivaldi-ffmpeg-codecs` package with a libffmpeg.so built from the Chromium source and makes Vivaldi load that library.

The build steps are based on the [AUR package with the same name](https://aur.archlinux.org/packages/vivaldi-ffmpeg-codecs/).

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

lluchs committed:
vivaldi-ffmpeg-codecs: init at 59.0.3071.104

Due to licensing costs, Vivaldi bundles a version of ffmpeg compiled
without support for the common H.264 codec. However, it is possible to
supply a custom libffmpeg.so with additional codecs. This derivation
uses the Chromium source to compile a compatible libffmpeg.so.

This approach is recommended by a Vivaldi developer, see
https://gist.github.com/ruario/bec42d156d30affef655

lluchs committed:
vivaldi: add proprietaryCodecs option

Fix #26413

volth commented on pkgs/applications/networking/browsers/vivaldi/ffmpeg-codecs.nix:
```diff
@@ -0,0 +1,51 @@
+stdenv.mkDerivation rec {
+  name = "${product}-${version}";
+  product = "vivaldi-ffmpeg-codecs";
+  version = "59.0.3071.104";
```
It might have sense to pick the ```version``` (or ```src```) from ```chromium``` package to have them in sync.

lluchs replied:
The Chromium version has to match the one Vivaldi is using, so it should be updated in sync with the `vivaldi` package, not with the `chromium` one.

volth replied:
then maybe
```
preBuild = ''
  if ! grep -F ${version} ${vivaldi}/opt/vivaldi/vivaldi-bin; then 
    echo "version mismatch"
    exit 1
  fi
'';
```
?

lluchs replied:
Wouldn't that introduce a circular dependency between `vivaldi` and `vivaldi-ffmpeg-codecs`?

In any case, this doesn't seem to work. I don't think it's possible to find the Chromium version without starting the browser and looking at vivaldi://about/ or checking the changelog.

volth replied:
Circular dependency is solvable, at least there is a trivial solution - put them into a single project.

The versions have good chance to diverge.

Browsers are updated often, usually in few hours after a new version is released, and those updates are make each time by different people. I do not expect all of the updaters would notice existence of ```vivaldi-ffmpeg-codecs``` and get an idea to run vivaldi, open vivaldi://about/ and update chrome version in ```vivaldi-ffmpeg-codecs```.

lluchs replied:
I just noticed that the AUR package includes a check for the Chromium version and added that to the package. This also made me notice that the currently packaged Vivaldi version isn't the latest. Thanks for the feedback!

0xABAB replied:
I would like to see the single project solution, or any other solution which cannot cause sync problems.

lluchs replied:
It's not possible for the packages to go out of sync now, the vivaldi build will fail if vivaldi-ffmpeg-codecs isn't built from the same Chromium version.

A single project for both would probably mean rebuilding ffmpeg even when Vivaldi didn't update its Chromium version. As the Chromium source is pretty large (~3.5 GB) and compiles rather slowly, I'd like to avoid that.

volth replied:
As ```proprietaryCodecs ? true``` it will be compiled anyway. Either on user machine or on ```hydra.nixos.org```

lluchs replied:
Yeah, but it shouldn't have to be compiled if Vivaldi updates without changing the underlying Chromium version.

0xABAB replied:
By out of sync I was talking about the case when a maintainer will forget about it and confusion will result. Any solution (for example adding a maintainer script) that solves that is what I would like to see. 

A secondary goal is efficiency of the amount of CPU being used.

lluchs committed:
vivaldi: 1.10.867.38-1 -> 1.10.867.48-1

volth commented on pkgs/applications/networking/browsers/vivaldi/default.nix:
```diff
@@ -40,7 +40,18 @@ stdenv.mkDerivation rec {
+  '';
+
   buildPhase = ''
+    runHook preBuild
```
Don't the hooks run authomatically?

lluchs replied:
Not when overriding buildPhase. [There's a couple packages doing the same thing.](https://github.com/NixOS/nixpkgs/search?utf8=✓&q=runhook+prebuild&type=)

volth replied:
then maybe
```nix
buildPhase = stdenv.lib.optionalString proprietaryCodecs ''
  check code
'' ++ ''
  actual build
'';
```
just to minimize the code a bit

lluchs replied:
I'd like to be able to override the check. It's very unpleasant to compile vivaldi-ffmpeg-codecs (extracting the 3.5 GB of code and building the gn tools takes ages) and the ffmpeg library doesn't seem to change too often.

lluchs committed:
vivaldi: fix file dialog crash

When trying to open or save a file using the file chooser GUI, Vivaldi
would crash with the message

   GLib-GIO-ERROR **: Settings schema 'org.gtk.Settings.FileChooser' is
   not installed

This commit adds the GTK directory to XDG_DATA_DIRS which fixes the
crash.

lluchs committed:
vivaldi: Add update script

lluchs commented:
I updated the PR. There's now an update script which will check the Vivaldi website for the current version and update both the `vivaldi` and `vivaldi-ffmpeg-codecs` packages. I removed the version check during the `vivaldi` build though. At the moment, Vivaldi uses Chromium version 59.0.3071.112, but there is no Chromium source tarball for that version. It's up to the maintainer to decide which Chromium version is best in that case. Currently, the previous version works fine.

I also included a (somewhat unrelated) patch for #27299.

lluchs commented:
Any further comments? Can someone merge this?

cc @otwieracz @nequissimus @volth @romildo

volth commented:
I cannot, from this list only @NeQuissimus should have enough access permissions
